### PR TITLE
Error refactor

### LINKED ISSUE #5: Better errors from procedural macro
There should be new APIs to do more than just panic when we encounter an error during codegen.

### PR DESCRIPTION
**Pull request by cmyr**
This is a bit of a mess, my apologies, in any case: this adds in some real error types, and separates out compile-time errors from runtime errors.

This also refactors pars-fmt; what was previously in `lib.rs` is now in `simple.rs`, except for error stuff, which is in `error.rs`. At some point in the future I expect to rename `pars-fmt` to `pars-internals`, and move more or less everything from `pars` into that.

closes #5

### CODE REVIEW

File: pars-fmt/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+use std::fmt;
+use std::ops::Range;
+
+/// An error that can occur when parsing format strings.
+/// This is a convenience; these errors will always occur at
+/// compile time, and this lets us give better diagnostics.
+#[doc(hidden)]
+#[derive(Debug, PartialEq)]
+pub struct FormatError {
+    inp: String,
```
Comment (gosulliv) [inline pars-fmt/src/error.rs]: This could use a better name. `input`?

File: pars-fmt/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
+//! Error types and related helpers
+
+use std::fmt;
+use std::ops::Range;
+
+/// An error that can occur when parsing format strings.
```
Comment (gosulliv) [inline pars-fmt/src/error.rs:6]: Good doc comment!

File: pars-fmt/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+}
+
+impl<'a> fmt::Display for MatchError<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MatchError::InputExhausted => write!(f, "input exhausted"),
+            MatchError::IncorrectFields { .. } => write!(f, "incorrect fields"),
+            MatchError::MissingSeparator { idx, string } => {
+                write!(f, "missing separator at index {}, text: '{}'", idx, string)
```
Comment (gosulliv) [inline pars-fmt/src/error.rs:69]: If we wanted to be fancy we could print an arrow like we do with `FormatError`.

File: pars-fmt/src/lib.rs [hunk 4]
```diff
@@ -1,290 +1,8 @@
...
-//! #[pars::fmt("#{x} #{y} #{width} #{height}")]
-//! struct Rect {
-//!     x: usize,
-//!     y: usize,
-//!     width: usize,
-//!     height: usize,
-//! }
-//!
-//!```
+//! Implementation for pars functionality. This lives in its own crate
```
Comment (gosulliv) [inline pars-fmt/src/lib.rs:1]: `Implementation for pars non-regex string parsing`?

File: pars-fmt/src/simple.rs [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    source: &'a str,
+    lead_separator: Option<Range<usize>>,
+    //NOTE: we represent all substrings as ranges of self.source, to avoid
+    //any unnecessary allocation.
+    fmt_fields: Vec<(Range<usize>, Range<usize>)>,
+    fields: Vec<String>,
+}
+
+#[allow(dead_code)]
+pub struct FmtMatch<'a, 'b> {
```
Comment (gosulliv) [inline pars-fmt/src/simple.rs:189]: I like this type. This is much better than returning a pair of lists.

File: pars-fmt/src/simple.rs [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+pub struct FmtMatcher<'a> {
+    source: &'a str,
+    lead_separator: Option<Range<usize>>,
+    //NOTE: we represent all substrings as ranges of self.source, to avoid
+    //any unnecessary allocation.
+    fmt_fields: Vec<(Range<usize>, Range<usize>)>,
+    fields: Vec<String>,
+}
+
+#[allow(dead_code)]
```
Comment (gosulliv) [inline pars-fmt/src/simple.rs:188]: I'd like to get rid of as many warning suppressions as we reasonably can; is this still needed?

**Review by gosulliv — APPROVED** [review]
Looks great!

File: pars-fmt/src/lib.rs (hunk 4 above)
Comment (cmyr) [reply pars-fmt/src/lib.rs:1]: yea, getting docs updated is a future todo

File: pars-fmt/src/simple.rs (hunk 5 above)
Comment (cmyr) [reply pars-fmt/src/simple.rs:189]: yea, and I think in upcoming work we can likely get rid of it completely, not totally sure yet though.

File: pars-fmt/src/simple.rs (hunk 6 above)
Comment (cmyr) [reply pars-fmt/src/simple.rs:188]: We can remove these after #12

File: pars-fmt/src/error.rs (hunk 3 above)
Comment (cmyr) [reply pars-fmt/src/error.rs:69]: we could, but unlike format error we don't expect this to always be displayed to the user? However yes, if they unwrap we could probably offer a better diagnostic, I'll add a todo.

**Pull request merged**